Add in a json parser so that the IDE can provide services around json editing.

CyrusNajmabadi opened the pull request:
Followup to https://github.com/dotnet/roslyn/pull/23984

So JFK has busted pipes.  So i've been in airport limbo :-(     On a good note, it gave me some time to expand on my previous Regex PR.  Now, on top of Regex support in the IDE we also support JSON strings.

This also expands the previous interaction model with built in json detection for string tokens.  i.e., if we see:

![image](https://user-images.githubusercontent.com/4564579/34694351-a0eb7324-f47b-11e7-8bfb-2a8f96b12d97.png)

Then we'll offer:

![image](https://user-images.githubusercontent.com/4564579/34697518-5a0dc490-f488-11e7-8bfd-bc76633d6711.png)

This then helps users see how they can add a comment that lights-up IDE features for these types of strings.  Note: i don't do this for regexes because of a concern about too many false positives.  Once the comment is added things like up like so:

![image](https://user-images.githubusercontent.com/4564579/34694398-d37a0684-f47b-11e7-9091-7ad16db391b3.png)

Of course, squiggles and braces are supported, on top of just colors:

![image](https://user-images.githubusercontent.com/4564579/34694425-eadd0132-f47b-11e7-8ee0-9ffffa5f69dc.png)

Also, if you're using JToken/JObject/JArray, then you don't need to provide the comment (similar to how ```new Regex(pattern)``` is automatically detected: 

![image](https://user-images.githubusercontent.com/4564579/34694784-33f2833c-f47d-11e7-86d0-a6f6a2578d00.png)

Things to note:

1. unlike regexes, this was much simpler to provide.  That's because json is *much* simpler and much better documented versus the .net regex model.
2. Two forms of json are supported.  A strict mode that aligns with the IETF rfc for json, as well as Json.net's format.  Json.net is clearly the .net json winner (almost 100m downloads) and is most likely what people will be using.  Json.net allows a superset of standard json strings (including allowing things like comments, as well as other constructs that would normally be errors in es6 json).  
3. This leverages the VirtualChar work from the regex work, demonstrating that this is a suitable subsystem for general DSL plugins.

CyrusNajmabadi commented:
Tagging @dotnet/roslyn-ide @Pilchie

jasonmalinowski commented:
@CyrusNajmabadi Should we wait until the regex one is resolved prior to looking at this one? This PR is quite, quite large.

CyrusNajmabadi commented:
> Should we wait until the regex one is resolved prior to looking at this one?

That woudl be my recommendation :)   

> This PR is quite, quite large.

It's a mere 46k lines.  If you do 2k lines an hour, you can finish it by tomorrow!

CyrusNajmabadi commented:
Options look like this:

![image](https://user-images.githubusercontent.com/4564579/34696339-457c125c-f483-11e7-8df4-295703d6ab56.png)

yaakov-h commented:
Can we just keep you in various airports? 😆

CyrusNajmabadi commented:
Tomorrow LaGuardia!

CyrusNajmabadi commented:
I have updated this PR to only be the json parser work. PR for IDE features around json strings has now moved into https://github.com/dotnet/roslyn/pull/25518

jcouv commented on src/Workspaces/Core/Portable/EmbeddedLanguages/Json/JsonHelpers.cs:
```diff
@@ -0,0 +1,36 @@
+             ImmutableArray<JsonTrivia> trailingTrivia, ImmutableArray<EmbeddedDiagnostic> diagnostics)
+            => CreateToken(kind, leadingTrivia, virtualChars, trailingTrivia, diagnostics, value: null);
+
+        public static JsonToken CreateToken(JsonKind kind,
```
This method is only used once (on line 20). Inline?

CyrusNajmabadi replied:
done.  #resolved.

jcouv commented on src/Workspaces/Core/Portable/EmbeddedLanguages/Json/JsonHelpers.cs:
```diff
@@ -0,0 +1,36 @@
+
+    internal static class JsonHelpers
+    {
+        public static JsonToken CreateToken(JsonKind kind, ImmutableArray<JsonTrivia> leadingTrivia, ImmutableArray<VirtualChar> virtualChars, ImmutableArray<JsonTrivia> trailingTrivia)
```
Nit: maybe break this line

CyrusNajmabadi replied:
done.  #resolved.

jcouv commented:
Could you retarget to the feature branch?

jcouv commented on src/Workspaces/Core/Portable/EmbeddedLanguages/Json/JsonKind.cs:
```diff
@@ -0,0 +1,48 @@
+{
+    internal enum JsonKind
+    {
+        None,
```
Is this used? If not, maybe `None = 0` or remove?

CyrusNajmabadi replied:
i like this as it makes it easy to test if a token is default.

jcouv replied:
Should it be `None = 0` then?

CyrusNajmabadi replied:
Sure.   i can do that.

CyrusNajmabadi replied:
#resolved.

jcouv commented on src/Workspaces/Core/Portable/EmbeddedLanguages/Json/JsonLexer.cs:
```diff
@@ -0,0 +1,413 @@
+            Text = text;
+        }
+
+        public VirtualChar CurrentChar => Position < Text.Length ? Text[Position] : new VirtualChar((char)0, default);
```
Nit: `span: default`

CyrusNajmabadi replied:
done.  #resolved.

jcouv commented on src/Workspaces/Core/Portable/EmbeddedLanguages/Json/JsonLexer.cs:
```diff
@@ -0,0 +1,413 @@
+
+        public VirtualChar CurrentChar => Position < Text.Length ? Text[Position] : new VirtualChar((char)0, default);
+
+        public ImmutableArray<VirtualChar> GetSubPattern(int start, int end)
```
`end` is generally `Position`. Consider adding an overload where end is implicit.

CyrusNajmabadi replied:
done.  #resolved.

jcouv commented on src/Workspaces/Core/Portable/EmbeddedLanguages/Json/JsonLexer.cs:
```diff
@@ -0,0 +1,413 @@
+                case '\'': case '"':
+                    return ScanString();
+
+                //case '-': case '.':
```
Why is this commented out?

CyrusNajmabadi replied:
Added comment explaining why we do not explicitly lex out numbers here, but instead let the parser take control.  #resolved.

jcouv commented on src/Workspaces/Core/Portable/EmbeddedLanguages/Json/JsonLexer.cs:
```diff
@@ -0,0 +1,413 @@
+        private (ImmutableArray<VirtualChar>, JsonKind, EmbeddedDiagnostic?) ScanString()
+        {
+            var start = Position;
+            var startChar = this.CurrentChar.Char;
```
Perhaps rename `openChar`?

CyrusNajmabadi replied:
done.  #resolved.

jcouv commented on src/Workspaces/Core/Portable/EmbeddedLanguages/Json/JsonLexer.cs:
```diff
@@ -0,0 +1,413 @@
+            return (chars, JsonKind.StringToken, diagnostic);
+        }
+
+        private EmbeddedDiagnostic? ScanEscape(int stringStart, int escapeStart)
```
`ScanEscape` could use a comment since unlike other scan methods, it doesn't return tokens or nodes, it just advances the position.

CyrusNajmabadi replied:
done.  #Resolved

jcouv commented on src/Workspaces/Core/Portable/EmbeddedLanguages/Json/JsonLexer.cs:
```diff
@@ -0,0 +1,413 @@
+            return (GetSubPattern(start, Position), JsonKind.TextToken, null);
+        }
+
+        private (ImmutableArray<VirtualChar>, JsonKind, EmbeddedDiagnostic?)? TryScanText(
```
This method seems unused

CyrusNajmabadi replied:
indeed.  removed.  #resolved.

jcouv commented on src/Workspaces/Core/Portable/EmbeddedLanguages/Json/JsonLexer.cs:
```diff
@@ -0,0 +1,413 @@
+            var chars = GetSubPattern(start, Position);
+            if (Position == start + 2)
+            {
+                var diagnostics = ImmutableArray.Create(new EmbeddedDiagnostic(
```
What if `//` are the last two characters in the file?

CyrusNajmabadi replied:
then you get an unterminated comment diagnostic.  This matches json.net.  You can see a test for that here: https://github.com/dotnet/roslyn/pull/25521/files#diff-d0475d1c64e0a4bb26665ae88ff3cf24R158

Will add a comment explaining this is intentional.

CyrusNajmabadi replied:
added comment.  #resolved.

jcouv commented on src/Workspaces/Core/Portable/EmbeddedLanguages/Json/JsonLexer.cs:
```diff
@@ -0,0 +1,413 @@
+                    GetTextSpan(start, Position))));
+        }
+
+        public TextSpan GetTextSpan(int startInclusive, int endExclusive)
```
This method seems to only be used once. private?

CyrusNajmabadi replied:
made privat.e  #resolved

jcouv commented on src/Workspaces/Core/Portable/EmbeddedLanguages/Json/JsonLexer.cs:
```diff
@@ -0,0 +1,413 @@
+        public TextSpan GetTextSpan(int startInclusive, int endExclusive)
+            => TextSpan.FromBounds(Text[startInclusive].Span.Start, Text[endExclusive - 1].Span.End);
+
+        public bool IsAt(string val)
```
private too?

CyrusNajmabadi replied:
done.  #resolved.

jcouv commented on src/Workspaces/Core/Portable/EmbeddedLanguages/Json/JsonLexer.cs:
```diff
@@ -0,0 +1,413 @@
+        public bool IsAt(string val)
+            => TextAt(this.Position, val);
+
+        private bool TextAt(int position, string val)
```
This method is only used in `IsAt`. Could be inlined?

CyrusNajmabadi replied:
I'm ok with it as is.  #won'tfix

jcouv commented on src/Workspaces/Core/Portable/EmbeddedLanguages/Json/JsonNodes.cs:
```diff
@@ -0,0 +1,342 @@
+        public JsonCompilationUnit(JsonSequenceNode sequence, JsonToken endOfFileToken)
+            : base(JsonKind.CompilationUnit)
+        {
+            Debug.Assert(sequence != null);
```
What's the sequence for an empty string?

jcouv commented on src/Workspaces/Core/Portable/EmbeddedLanguages/Json/JsonNodes.cs:
```diff
@@ -0,0 +1,342 @@
+    }
+
+    /// <summary>
+    /// Represents a chunk of text (usually just a single char) from the original pattern.
```
Wrong comment?

CyrusNajmabadi replied:
fixed.  #resolved.

jcouv commented on src/Workspaces/Core/Portable/EmbeddedLanguages/Json/JsonNodes.cs:
```diff
@@ -0,0 +1,342 @@
+    internal sealed class JsonNegativeLiteralNode : JsonValueNode
+    {
+        public JsonNegativeLiteralNode(JsonToken minusToken, JsonToken literalToken)
+       : base(JsonKind.NegativeLiteral)
```
Nit: formatting

CyrusNajmabadi replied:
fixed.  #resolved.

jcouv commented on src/Workspaces/Core/Portable/EmbeddedLanguages/Json/JsonParser.cs:
```diff
@@ -0,0 +1,476 @@
+                _lexer.Text, root, diagnostics);
+        }
+
+        private EmbeddedDiagnostic? CheckTopLevel(
```
static?

CyrusNajmabadi replied:
done.  #resolved.

jcouv commented on src/Workspaces/Core/Portable/EmbeddedLanguages/Json/JsonParser.cs:
```diff
@@ -0,0 +1,476 @@
+
+        private static JsonToken GetFirstToken(JsonNode node)
+        {
+            foreach (var child in node)
```
Why do we need `foreach`?

CyrusNajmabadi replied:
indeed. #resolved.

jcouv commented on src/Workspaces/Core/Portable/EmbeddedLanguages/Json/JsonParser.cs:
```diff
@@ -0,0 +1,476 @@
+            throw new InvalidOperationException();
+        }
+
+        private static EmbeddedDiagnostic? GetDiagnostic(JsonNode node)
```
A few of the overloads could be renamed `GetFirstDiagnostic`.

CyrusNajmabadi replied:
done.  #resolved

jcouv commented on src/Workspaces/Core/Portable/EmbeddedLanguages/Json/JsonParser.cs:
```diff
@@ -0,0 +1,476 @@
+        {
+            var list = ArrayBuilder<JsonValueNode>.GetInstance();
+
+            if (ShouldConsumeSequenceElement())
```
`while` loop would be fine

CyrusNajmabadi replied:
done.  #resolved

jcouv commented on src/Workspaces/Core/Portable/EmbeddedLanguages/Json/JsonParser.cs:
```diff
@@ -0,0 +1,476 @@
+
+        private JsonValueNode ParseLiteralOrPropertyOrConstructor()
+        {
+            // var token = ConsumeCurrentToken().With(kind: JsonKind.TextToken);
```
Dead comment

CyrusNajmabadi replied:
done.  #resolved.

jcouv commented on src/Workspaces/Core/Portable/EmbeddedLanguages/Json/JsonParser.cs:
```diff
@@ -0,0 +1,476 @@
+
+        private JsonObjectNode ParseObject()
+        {
+            var savedInObject = _inObject;
```
This seems wrong, and I'm surprised this would not cause bugs.
I'd expect that we're in a single context at a given time (either object, array or constructor). But it looks like this invariant is not protected.
I would have thought the other two properties would have been saved and reset to `false`. 
Alternatively, I think we should use an enum for context (None, Object, Array, Constructor).

CyrusNajmabadi replied:
> I'd expect that we're in a single context at a given time (either object, array or constructor). 

Nope.  This works like how things work in the C#/TS parsers.  You have a stack of contexts that you're in, and you're pushing/popping from that stack.  However, for efficiency, that stack is stored as an int in C#/TS, and i just used 3 bools here.

THe reason you don't actually need an unbounded stack is that you only need to keep track if you're in some type off construct the first time you enter it.  If you recursively enter that same construct again, you don't actually need to record that.  

Then, this information is simply used to know what you should do if you encounter something unexpected during parsing.  For example, if you see.    ```{ ... [ ... }```, i.e. we see a close brace, while in the middle of parsing an array.  Because we know we're *also* in an object, we know we need to bail out of parsing the array so that the object we're parsing can consume hte close brace.

CyrusNajmabadi commented:
@jcouv Can you make me a branch called features/embeddedJson  ?  it should be based of of features/embeddedRegex  thanks!

jcouv commented on src/Workspaces/Core/Portable/EmbeddedLanguages/Json/JsonParser.cs:
```diff
@@ -0,0 +1,476 @@
+                return false;
+            }
+
+            for (int i = 0, n = val.Length; i < n; i++)
```
I don't think there is a need to save `val.Length` into a local. I think Vlad said the JIT recognizes this pattern (doesn't recompute Length on every iteration).

CyrusNajmabadi replied:
Wow.  They can do this for immutable array?  that's pretty impressive :)  I will change.

CyrusNajmabadi replied:
oh wait. this is a string.  of course...

jcouv commented on src/Workspaces/Core/Portable/EmbeddedLanguages/Json/JsonParser.cs:
```diff
@@ -0,0 +1,476 @@
+            return true;
+        }
+
+        private bool IsDigit(char ch)
```
static?
Also applies to a few methods below (ParseLIteral, ParseNumber, ParseEmptyValue).

CyrusNajmabadi replied:
done.  #resolved.

jcouv commented on src/Workspaces/Core/Portable/EmbeddedLanguages/Json/JsonParser.JsonNetSyntaxChecks.cs:
```diff
@@ -0,0 +1,237 @@
+
+    internal partial struct JsonParser
+    {
+        private partial struct JsonNetSyntaxChecker
```
Why partial?

CyrusNajmabadi replied:
fixed.  #resolved.

jcouv replied:
Is there a spec or doc we could point to for both sets of rules (strict and Json.Net)?

jcouv replied:
Can we make all the methods here static? (same for the other checker)

CyrusNajmabadi replied:
done.  also made them into static classes.

CyrusNajmabadi commented:
> Is there a spec or doc we could point to for both sets of rules (strict and Json.Net)?

For json.net, no.  You have to look at its impl.

For strict, yes.  It's:  https://tools.ietf.org/html/rfc8259

jcouv commented on src/Workspaces/Core/Portable/EmbeddedLanguages/Json/JsonParser.cs:
```diff
@@ -0,0 +1,476 @@
+        }
+
+        private JsonEmptyValueNode ParseEmptyValue()
+            => new JsonEmptyValueNode(ConsumeCurrentToken());
```
Why is this called "empty value"?

CyrusNajmabadi replied:
it's used for cases like ```{ a: , }``` or ```[,,,]```  Perhaps 'OmittedValue'?

jcouv commented on src/Workspaces/Core/Portable/EmbeddedLanguages/Json/JsonParser.JsonNetSyntaxChecks.cs:
```diff
@@ -0,0 +1,237 @@
+            }
+
+            private static bool IsLegalPropertyNameChar(char ch)
+                => char.IsLetterOrDigit(ch) | ch == '_' || ch == '$';
```
Did you mean || ?

CyrusNajmabadi replied:
whoops!

CyrusNajmabadi replied:
fixed.  #resolved

CyrusNajmabadi commented:
Note: An issue of interest is: https://github.com/JamesNK/Newtonsoft.Json/issues/646

This is a request of json.net to have a strict mode.  I communicated with James on that issue and we enumerated all the ways we think that json.net deviates from the standard.  I have included code and tests for all those cases.  From that bug, the list is:

-     Properties with single quotes instead of double quotes
-     Properties without quotes
-     NaN, Infinity, -Infinity
-     JavaScript constructors, e.g. new DateTime(12233434)
-     Single and multi-line comments
-     Trailing commas
-     Empty commas, e.g. [,,,]
-     Octal numbers (number starts with a zero)
-     Hexadecimal numbers (I think...)
-     New lines in strings
-     undefined
-     lenient number parsing
-     allows more whitespace than strict

CyrusNajmabadi commented:
> I also think it would be good to squash some of those 250+ commits ;-)

There are so many branches bouncing around, with code cross pollination happening between all of them.  I'm terrified of what will happen if i start squashing, and then trying to make things flow properly :-/

Neme12 commented:
Why not just 1 branch for both embedded languages?

CyrusNajmabadi commented:
> Why not just 1 branch for both embedded languages?

Because they do not depend on each other.  They are independent features.  It also means they can be merged in independently.

CyrusNajmabadi commented:
@jcouv I've rebased this to the new branch.

CyrusNajmabadi commented on src/Workspaces/CSharpTest/CSharpServicesTest.csproj:
```diff
@@ -30,6 +30,8 @@
     <Reference Include="System" />
     <Reference Include="System.Core" />
+    <Reference Include="System.Runtime.Serialization" />
+    <Reference Include="System.Web.Extensions" />
```
intentional.  each of these has a json parser that we reference in order to validate our own parser against.

CyrusNajmabadi commented:
@jinujoseph Is there someone who could help out here and be a second reviewer for the IDE side?  Julien is already helping out and reviewing all the regex/json PRs.  It would be nice if we could get a second pair of eyes so this could go in sometime soon.

CyrusNajmabadi commented:
Also, in case it matters, the json work is *much* easier to review than the regex work.   Parsing regexes is *complex* work (especially when trying to match what .net does).  Json, in comparison is trivial.  So having someone review the json work first would go faster, and then help prime them for a better understanding when going through the regex work.

jinujoseph commented:
Sure ,  looping in @sharwell for review

CyrusNajmabadi commented:
retest this please.

CyrusNajmabadi commented:
Thanks @jinujoseph !   @sharwell , let me know if there's any information i can provide to help out here.  Of interest may be the previous two PRs:

1. https://github.com/dotnet/roslyn/pull/24111.  This PR adds in the subsytem that allows one to map between the actual 'Text' chars written in source, to the interpretted 'ValueText' chars that are the ones processed by an embedded language like Regex of Json.
2. https://github.com/dotnet/roslyn/pull/25541.  This PR adds in a bunch of common code around embedded syntax that are then used to prevent lots of duplication when i did the Regex and Json embedded languages.

CyrusNajmabadi commented:
Also, this is all being done in a branch so that things can be nicely separated into independent PRs, with the final work going in at once.

fanoI commented:
It will test be a public API that will permit to add new "embedded languages" if we want?
This will permit us to cover: https://github.com/dotnet/csharplang/issues/226

CyrusNajmabadi commented:
> It will be a public API that will permit to add new "embedded languages" if we want? This will permit us to cover: dotnet/csharplang#226

My approach is about allowing language-services to light-up inside C#/VB strings when they contain a snippet of some other language.  So, if you wanted to do inline-IL through C#/VB string-literals, then this could be utilized.

>  It will be a public API 

It is not intended to be a public API initially.  I'm sure the roslyn team can determine if that fits their goals and make it public once it has received enough time and has proved itself.

CyrusNajmabadi commented:
@jcouv Can you let me know if there are any other changes you'd like to see made?  @jmarolf You might also be interested in reviewing this.

jcouv commented on src/Workspaces/Core/Portable/EmbeddedLanguages/Json/JsonOptions.cs:
```diff
@@ -0,0 +1,50 @@
+{
+    internal class JsonOptions
+    {
+        public static PerLanguageOption<bool> ColorizeJsonPatterns =
```
Could you share screenshots?
Also, are we sure those settings should be per-language (C# vs. VB)?

CyrusNajmabadi replied:
Screenshots of what?  Colored json?  If so, it's at the top of hte file.  And the settings should be per-language.  People should be allowed to say "i want to turn on embedded json in C# files, but not in VB files" (or vice versa).  

Options we expose are almost always per-language so that they can show up in tools|options and be independently flipped in the C# or VB sections.  We have some exceptiosn (like "full solution analysis"), but that's because we don't have a good system for even defining what it means for that setting to be single-language :)

jcouv replied:
I meant the settings page  ;-)

CyrusNajmabadi replied:
Gotcha.  Looks like this:

![image](https://user-images.githubusercontent.com/4564579/38476021-7ef54f52-3b60-11e8-8086-9f86a466aa4f.png)

CyrusNajmabadi commented:
@jmarolf @sharwell Do you want to take a look?  Json parsing is actually pretty basic (esp. compared to regex).  So it'll be a lot easier to review this PR to get a sense for how all these new components work.

Thanks!

CyrusNajmabadi commented:
@dotnet/roslyn-ide can someone assist @jcouv with this review?  This is a pretty simple json parser that supports json.net style json parsing as well as strict rfc8259 parsing.  It will be used by hte json-language-services feature to actually do the parsing.

CyrusNajmabadi commented:
@jinujoseph @sharwell @kuhlenh Can you let me know a timeline here for when this might get attention?  Last involvement from teh team was 5 weeks ago.  I've done a bunch to try to make these PRs easier to review and get into a feature branch, but i simply don't know if these are on anyone's radar or when i might hope they can get looked at.

Thanks!

Neme12 commented:
There's a ton of changes from master here.

CyrusNajmabadi commented:
> There's a ton of changes from master here

I'm right in the middle of updating the feature branches and all downstream branches unfortunately.

justcla commented:
FYI: JetBrains have implemented support for the short form "lang=" (as well as the original "language=" form).
It will be available in ReSharper 2018.2

CyrusNajmabadi commented:
> FYI: JetBrains have implemented support for the short form "lang=" (as well as the original "language=" form).

Were you asking me to support both those forms?  If so, i already do :)  

Also, would you be able to assist in any way with teh reviews on this.  It looks like this has completely stalled out and isn't making any forward progress, even after 6 months...

justcla commented:
When reviewing it, I noticed that you implemented lang= and language=.
I noticed that JetBrains also implemented as similar behavior, but they were only using the language= form. So that people using ReSharper would get the behavior expected by users who used your short form, I asked JetBrains if they would consider also implementing the short form.
And I just found out that they did. (Thanks JetBrains!)
So that makes it easier to bring your code into the product - as we will have consistency among users with different tools.

As for progress, it's just a matter of resourcing. The team has been prioritizing other features for 15.8. (You know how it is.)

CyrusNajmabadi commented:
@sharwell @jinujoseph How would you like to proceed on this?

CyrusNajmabadi commented:
@sharwell @jinujoseph How would you like to proceed on this?

jinujoseph commented:
Unfortunately, This feature is currently not in the top priority for us to pursue, so this is on hold at this moment. Following up to see if we can get merge this to feature branch

jinujoseph commented:
@CyrusNajmabadi as discussed closing this PR for now.
